libyul/optimiser/LoadResolver.cpp

```diff
@@ -64,6 +65,22 @@ void LoadResolver::visit(Expression& _e)
 				tryResolve(_e, location, funCall->arguments);
 				break;
 			}
+		if (
+			!m_containsMSize &&
+			toEVMInstruction(m_dialect, funCall->functionName.name) == evmasm::Instruction::KECCAK256
+		)
+		{
+			Identifier const* start = get_if<Identifier>(&funCall->arguments.at(0));
+			Identifier const* length = get_if<Identifier>(&funCall->arguments.at(1));
+			if (start && length)
+				if (auto value = util::valueOrNullptr(m_keccak, make_pair(start->name, length->name)))
+					if (inScope(*value))
```

Comment:
Oh, so `DataflowAnalyzer` does not clear the variables after exiting a scope/block? Is this for a situation this?
```solidity
{
    let a := 42;
    let b := 42;
    sstore(0, keccak256(a, b))
}
{
    let x := 42;
    let y := 42;
    sstore(0, keccak256(x, y))
}
```
If not, can you add a test that actually fails this condition?

Reply:
The problem is in the value:
```
let a := 0
let b := 32
{
  let c := keccak(a, b)
}
let d := keccak(a, b)
```
We cannot replace the RHS of the last line by `c` because it is not in scope. The DataFlowAnalyzer might erase the knowledge, but I think this is good to be here, just to be sure.